### PR TITLE
Improve asynchronous host to device copy

### PR DESCRIPTION
**Pull request by okuta**
This PR implement asynchronous copy to `cupy.array`.
The stream option in `cuda.to_gpu` is deprecated, because stream option is difficult.

### CODE REVIEW

**Comment by gwtnb** [comment]
Please resolve conflicts.

**Comment by okuta** [comment]
I resolved conflicts/

File: cupy/cuda/pinned_memory.pyx [hunk 1]
```diff
@@ -129,12 +130,60 @@ cdef class PinnedMemoryPointer:
...
+        if event.done:
+            return
+        with self.lock:
+            self.events.append((event, obj))
+
+    cpdef check_and_release(self):
+        """ Check and release completed events.
+
+        """
+        if len(self.events) == 0:
```
Comment (gwtnb) [inline cupy/cuda/pinned_memory.pyx]: Is it safe to touch `self.events` without lock?

Comment (okuta) [inline cupy/cuda/pinned_memory.pyx]: Python has GIL...

**Comment by gwtnb** [comment]
How about like this?
```diff
diff --git a/cupy/cuda/pinned_memory.pyx b/cupy/cuda/pinned_memory.pyx
index 2b0d9bea..5587fda4 100644
--- a/cupy/cuda/pinned_memory.pyx
+++ b/cupy/cuda/pinned_memory.pyx
@@ -148,10 +148,10 @@ cdef class _EventWatcher:
             event (cupy.cuda.Event): The CUDA event to be monitored.
             obj: The object to be held.
         """
-        self.check_and_release()
-        if event.done:
-            return
         with self.lock:
+            self._check_and_release_without_lock()
+            if event.done:
+                return
             self.events.append((event, obj))
 
     cpdef check_and_release(self):
@@ -161,9 +161,11 @@ cdef class _EventWatcher:
         if len(self.events) == 0:
             return
         with self.lock:
-            while len(self.events) != 0 and self.events[0][0].done:
-                del self.events[0]
+            self._check_and_release_without_lock()
 
+    cpdef _check_and_release_without_lock(self):
+        while len(self.events) != 0 and self.events[0][0].done:
+            del self.events[0]
```

**Comment by okuta** [comment]
Thank you.
I will fix.

**Comment by okuta** [comment]
@gwtnb Cloud you merge this PR?

**Comment by okuta** [comment]
Please merge https://github.com/cupy/cupy/pull/123 before this PR.

**Comment by gwtnb** [comment]
LGTM

**Comment by kmaehashi** [comment]
> The stream option in cuda.to_gpu is deprecated, because stream option is difficult.

Are there any alternate ways to use CUDA stream with Chainer?

**Comment by okuta** [comment]
I have no idea about implementing safely `to_gpu` with CUDA stream.
You can use `copy_from_device_async` directory. That method understands CUDA stream option.

**Pull request merged**